x/wasm/keeper/keeper.go

```diff
@@ -1005,6 +1029,116 @@ func (k Keeper) newQueryHandler(ctx sdk.Context, contractAddress sdk.AccAddress)
 	return NewQueryHandler(ctx, k.wasmVMQueryHandler, contractAddress, k.gasRegister)
 }
 
+func (k Keeper) updateSizeForContract(ctx sdk.Context, contractAddress sdk.AccAddress, sizeChange int64) error {
+	// always charge rent first before updating size, as the charging logic assumes constant size since last
+	// charged block
+	if err := k.chargeRent(ctx, contractAddress); err != nil {
+		return err
+	}
+	sizeKey := types.GetContractStoreSizePrefix(contractAddress)
+	store := ctx.KVStore(k.storeKey)
+	currentSize, err := k.getCurrentSizeForContract(ctx, contractAddress)
+	if err != nil {
+		return types.ErrUpdateContractSize.Wrap(err.Error())
+	}
+	currentSizeSigned := int64(currentSize) + sizeChange
+	if currentSizeSigned < 0 {
+		// this should not happen
+		panic("negative size for contract")
+	}
+	newBz := make([]byte, 8)
+	binary.BigEndian.PutUint64(newBz, uint64(currentSizeSigned))
+	store.Set(sizeKey, newBz)
+	return nil
+}
+
+func (k Keeper) getCurrentSizeForContract(ctx sdk.Context, contractAddress sdk.AccAddress) (uint64, error) {
+	return GetSize(ctx, contractAddress, k.storeKey)
+}
+
+func (k Keeper) chargeRent(ctx sdk.Context, contractAddress sdk.AccAddress) error {
+	rentInfo, err := k.getRentInfo(ctx, contractAddress)
+	if err != nil {
+		return err
+	}
+	duration := uint64(ctx.BlockHeight()) - rentInfo.LastChargedBlock
+	if duration == 0 {
+		// shortcut in the case of same block changes
+		k.setRentInfo(ctx, contractAddress, rentInfo)
+		return nil
+	}
+	size, err := k.getCurrentSizeForContract(ctx, contractAddress)
+	if err != nil {
+		return err
+	}
+	if size == 0 {
+		rentInfo.LastChargedBlock = uint64(ctx.BlockHeight())
+		k.setRentInfo(ctx, contractAddress, rentInfo)
+		return nil
+	}
+	rentPrice := k.GetParams(ctx).UnitRentPrice
+	rentToBeCharged := sdk.NewDec(int64(duration * size)).Mul(rentPrice).RoundInt().Int64()
+	// protect against overflow
+	if rentInfo.Balance < rentToBeCharged+math.MinInt64 {
+		rentInfo.Balance = math.MinInt64
+	} else {
+		rentInfo.Balance -= rentToBeCharged
+	}
+	rentInfo.LastChargedBlock = uint64(ctx.BlockHeight())
+	k.setRentInfo(ctx, contractAddress, rentInfo)
+	if rentInfo.Balance < 0 {
```

Comment:
just to confirm, the expected behavior here is that if a contract has insufficient duration rent, then the contract is borked until the deposit rent tx is done, right? and this would result in any executes, migrates, etc to fail

Reply:
yeah that's correct